### ISSUE TITLE
Object destructuring

### ISSUE THREAD

**Issue by RamonGebben** [yoshuawuyts/es2020#1]
One of the features I use very often is Object destructuring.
It would be awesome if you could include this as well.

**Comment by yoshuawuyts** [comment]
Thanks for opening!

To be honest I find object destructuring to be quite confusing - with array destructuring being added perhaps even more so. I do see the point for it though, but I prefer to recommend relying on [xtend](https://github.com/Raynos/xtend) for now.

**Comment by yoshuawuyts** [comment]
Hmm, more people like object destructuring it seems. Keeping it closed for now, but creating a little backlog:
- https://github.com/bendrucker/bundleify/issues/2#issuecomment-221338269
- https://twitter.com/davidmarkclem/status/735155305290432512

**Comment by yoshuawuyts** [comment]
Someone is doing a great job convincing me here - https://twitter.com/davidmarkclem/status/735156452604547072

Not saying quite yes right now... but maybe. Reopening.

cc/ @davidmarkclements

**Comment by yoshuawuyts** [comment]
Also cc/ @ahdinosaur @bendrucker - what's your take on object destructuring? Worth the addition (never used it before myself)?

**Comment by bendrucker** [comment]
It's an interesting feature/syntax. I don't use it right now. My concern is that it's a much broader and more complex feature than templates, arrow functions, or `const`. At very least it's worth leaving it open for a couple of days to measure consensus.

**Comment by yoshuawuyts** [comment]
@bendrucker yeah, you're making a good point - it actually changes the language in a fundamental way. 

I'll leave it open for now, let's see how other people feel

**Comment by ajoslin** [comment]
As @bendrucker mentioned, it's a bloated feature that has its good and bad parts. Here's what I like and don't like:

:+1::
- Structuring of keys: 
  - `const key = 'key'; const object = {[key]: 1}`
  - `const value = 'foo'; const object = {value}`
- Destructuring of keys:
  - `var fn = ({foo, bar}) => ()`
  - `const {foo} = require('dependency')`

:-1::
- Destructuring of arrays: `const [one, two] = [1, 2]`
- `fn.apply` via spread: `array.push(...myValues)`
- Defaults via structuring: `var fn = ({foo = 3, bar = 4} = {}) => ()`
  - The `=` in objects is not even javascript, and the defaults only apply in very specific conditions.

So object (de)structuring is definitely a mixed bag of good and bad. The only way I would 'publicly recommend' it is if I could give someone a strict linter to go with it. But the complexity may not even be worth it when you're going that far.

**Comment by ahdinosaur** [comment]
i like destructuring, i [learned their power with clojure](http://clojure.org/reference/special_forms#binding-forms). echo @ajoslin for my use cases, except might like array destructuring (to return tuples from functions) but not sure. rest, spread, and defaults seem separate, and i'm less of a fan for these.

**Comment by chrisinajar** [comment]
I'll play the role of devils advocate here and say this is JavaScript and not Python. Sugar should be for safety and syntax, not for structures (or control flow.) 👎

**Comment by yoshuawuyts** [comment]
Hmmm, so I'm starting to gravitate towards a moderate "yes" on this. Even though like with most things in JS it can be abused, there are definitely some upsides. I think @ajoslin's analysis is pretty good, and if we document the good and the bad it could be included.

I mean: `const`'s block scoping arguably has no place in JS either, but the upsides of having mutability outweigh the double scoping rules. Perhaps if we treat destructuring the same it's not all that bad.

## See Also
- https://twitter.com/dam/status/735469726105632769
- https://twitter.com/markdalgleish/status/735232841563312128
- https://twitter.com/davidmarkclem/status/735154490395889664
- https://github.com/KoryNunn/es-twentysixnext

**Comment by davidmarkclements** [comment]
@ajoslin analysis is good, but as @ahdinosaur points out, it takes a broader look at "structure related syntax". In terms of _destructuring_ we can limit the discussion to the following:
- object destructuring 
- array destructuring
- destructuring via params
- destructuring via assignment

I'm +1 on all of these. 

Argument for array destructuring is tuples, and reference swapping without tmp vars (`[b, a] = [a, b]`).

Destructured property defaults should only be considered if default params are on the menu, seems like those things should be tied together. 

Multi-tier destructuring also should be considered, a full implementation would support that but perhaps es2020 may wish to avoid it for philosophical reasons, example of multi-tiered: 

```
const {a: {b}} = {a: {b: 2}}
console.log(b) //2
```

**Comment by TehShrike** [comment]
I love object destructuring in function arguments.  Named function arguments are vastly superior to ordered arguments in most cases, and with object destructuring, they suddenly become very low-friction in JS.

**Comment by mgol** [comment]
fn.apply via spread is extremely useful. In ES6 you don't need to use the broken `arguments` object at all. If, say, you want to invoke another function with the same arguments as the ones you received but just adding an additional parameter, it's easy now:

```
const f = (...params) => g(1, ...params)
```

Much more readable than the terribly imperative non-spread version (you also can't use arrows without rest parameters):

```
const f = function () {
    const args = Array.prototype.slice.call(arguments);
    args.unshift(1);
    return g.apply(null, args);
};
```

**Comment by kristoferjoseph** [comment]
_Moved to spread cuz I got em confused like a nimrod_

**Comment by KoryNunn** [comment]
Can you lock destructing assignment out of params? if so, then I'm +1.

Usage within params is horribly confusing.

**Comment by davidmarkclements** [comment]
@KoryNunn could that just be culture shock?

Param destructuring is what makes destructuring so appealing

**Comment by yoshuawuyts** [comment]
Alright, param destructuring is out - if an object isn't explicitely passed the function crashes and burns, which feels wrong. Sometimes it's ok for an options object not to exist, even if it has properties attached. There's ways around it, but it introduced a _lot_ of weirdness that wasn't there before it.

<img width="616" alt="screen shot 2016-05-27 at 23 34 18" src="https://cloud.githubusercontent.com/assets/2467194/15611086/a0ec888e-2463-11e6-9958-42372887d50b.png">

**edit:** to clarify: I don't want to be passing `{}` whenever object params are optional. I'm sure it's a rule that can be learned, but it feels like a quirk - agree with @korynunn here: it feels off

**edit:** it also doesn't work in Node prior to 6.0 - feel like we shouldn't be using it if it doesn't run on the latest LTS - don't want to be blowing up Node all over the place / having to rely on some `-r` flag on startup :/

**Comment by yoshuawuyts** [comment]
So this mostly leaves:
- Destructuring of keys: `const key = 'key'; const object = {[key]: 1}`
- Destructuring on assignment: `const vec = { x: 12, y: 14 }; const {x, y} = vec`

I'm not quite sure this is worth it, as the alternatives are very readable, despite being slightly more verbose.

fwiw I agree that array destructuring doesn't seem great - haven't seen a strong argument in favor of it yet.

---

RE: @kristoferjoseph that would be related to spread I reckon - https://github.com/yoshuawuyts/es2020/issues/3

**Comment by ajoslin** [comment]
To make the destructuring of defaults in function args work better, you can pass in a default value of empty object for the options:

But then we get into the weirdness of defaults only working against undefined, not null...

![](http://ajoslin.co/V1Bv/36vSuUEr+)

**Comment by yoshuawuyts** [comment]
@ajoslin ah, thanks for answering that - yeah, I think that makes things even _more_ confusing 😞

**Comment by kristoferjoseph** [comment]
@yoshuawuyts 💩 You are right! I put this in the wrong issue  
[😬](http://media.riffsy.com/videos/eab7bd66b9ae430aafc009ccc22bcc84/mp4)

**Comment by orodio** [comment]
My perspective on this may be a bit different. I was first introduced to a similar concept (pattern matching) in Haskell and Erlang where it is one of the most powerful features of the language and something I miss wholeheartedly in JS. The JS implementation therefore feels like a half-assed attempt that sort of misses the point to me (especially in the whole destructuring arguments paradigm).

Here is a completely contrived example in elixir of a function that is passed a map, causes an analytical side effect and returns the original map.

```
def track_user_update(user = %{ guid }) do
  analytical_event({ :user_update, guid })
  user
end
```

There are two main points to that snippet:
1 - I wasn't only left with the destructed parts, I still have the entire `user` map that was passed to it
2 - I have decoratively said that this function cares about the `guid`.

If we were to write this same thing in javascript we have a couple options

```
// 1
function track_user_update(user) {
  analytical_event({ type: "user_update", guid: user.guid })
  return user
}
```

In number one it is hard to tell that the thing i care about is the `guid`. We have lost that explicit nature. 

```
// 2
function track_user_update(user) {
  const { guid } = user
  analytical_event({ type: "user_update", guid:guid })
  return user
}
```

In number two it is more obvious that we care about the guid. But that line could have just as easily been written like `const guid = user.guid`

```
// 3
function track_user_update({ guid }) {
  analytical_event({ type:"user_update", guid:guid })
}
```

In number three we have lost our ability to return the original `user` that was passed in.

Where I see the destructuring really shining is when we care about more than two things. In our number two example it seems a little overkill for a single variable. But take this example:

```
function foo(bar, flop) {
  const { guid, baz, frog } = bar
  const { bing, bong, bam } = flop
  analytical_event({ type:"did_the_thing", guid:guid, baz:baz, frog:frog, bing:bing, bong:bong, bam:bam })
  return bar
}
```

 vs

```
function foo(bar, flop) {
  const guid = bar.guid
  const baz  = bar.baz
  const frog = bar.frog
  const bing = flop.bing
  const bong = flop.bong
  const bam = flop.bam
  analytical_event({ type:"did_the_thing", guid:guid, baz:baz, frog:frog, bing:bing, bong:bong, bam:bam })
  return bar
}
```

I find the first one tells me more about what i are about, mainly that we want the `guid` `baz` and `frog`, that they have a direct relationship to the `bar`, that we want the `bing` `bong` and `bam` and that they are related to the `flop`. You can see that in the second example too but you have to pay way more attention, i find it way less obvious at a glance, sure we could space things out a bit differently, could group them by only having two const declarations, but now we are relying on a made up convention to show meaning that is obvious in the first example.

TL;DR
i think it has some pros and cons, but can definitely be a useful tool for good.

**Comment by yoshuawuyts** [comment]
@orodio thanks for the big write up! 

I think this is how I'd do it in plain ES5:

``` js
function foo(bar, flop) {
  analytical_event({
    type: "did_the_thing",
    guid: bar.guid,
    baz: bar.baz,
    frog: bar.frog,
    bing: flop.bing,
    bong: flop.bong,
    bam: flop.bam
  })
  return bar
}
```

But yeah, destructuring + shorthand props definitely makes it possible to do it more efficiently:

``` js
function foo(bar, flop) {
  const { guid, baz, frog } = bar
  const { bing, bong, bam } = flop
  const ev = { type: 'did_the_thing', guid, baz, frog, bing, bong, bam }
  analytical_event(ev)
  return bar
}
```

I do think you're right and this communicates better what's going on. HMMMMM THIS IS TOUGH YOU PEOPLE, YOU ALL MAKE FINE ARGUMENTS

**Comment by yoshuawuyts** [comment]
So we've decided against this for `es2020`, but if you want it anyway consider using https://github.com/ahdinosaur/es2040 :sparkles:
